### PR TITLE
Remove invalid cancelled signals after retrying

### PR DESCRIPTION
**Pull request by huydhn**
This addresses a confusing bug on HUD and Dr.CI where a bunch of unrelated cancelled signals showing up, forcing people to force merge.  For example,

* Dr.CI https://github.com/pytorch/pytorch/pull/107339
* HUD https://hud.pytorch.org/pr/107339
* Dr.CI https://github.com/pytorch/pytorch/pull/105055
* HUD https://hud.pytorch.org/pr/105055 

They are cancelled signals from the previous workflow run that had been retried successfully.  The cancelled signal were there because the names are different, i.e. `manywheel-py3_10-cuda11_8-test (cancel)` became `manywheel-py3_10-cuda11_8-test / test (success)` after retrying

The fix I have here is to use a trie search to check if a cancelled job has been retried successfully and remove it from the list accordingly.

### Testing

* https://torchci-git-fork-huydhn-remove-wrong-cancel-948947-fbopensource.vercel.app/pr/107339
* https://torchci-git-fork-huydhn-remove-wrong-cancel-948947-fbopensource.vercel.app/pr/105055

* **Dr.CI  #107339**

### :test_tube: See artifacts and rendered test results at [hud.pytorch.org/pr/107339](https://hud.pytorch.org/pr/107339)
* :page_facing_up: Preview [Python docs built from this PR](https://docs-preview.pytorch.org/pytorch/pytorch/107339/index.html)
* :page_facing_up: Preview [C++ docs built from this PR](https://docs-preview.pytorch.org/pytorch/pytorch/107339/cppdocs/index.html)
* :question: Need help or want to give feedback on the CI? Visit the [bot commands wiki](https://github.com/pytorch/pytorch/wiki/Bot-commands) or our [office hours](https://github.com/pytorch/pytorch/wiki/Dev-Infra-Office-Hours)

Note: Links to docs will display an error until the docs builds have been completed.

## :x: 1 New Failure, 6 Unrelated Failures
As of commit 59b327cdb6891318111fe98c0dbc72c7da0e7b95 with merge base 5531a23b204b4daa2c0bb3c52610e9a0ba79dacf (<sub><sub>image</sub></sub>):
<details open><summary><b>NEW FAILURE</b> - The following job has failed:</summary><p>

* [wheel-py3_10-cpu-test](https://hud.pytorch.org/pr/pytorch/pytorch/107339#16727706927) ([gh](https://github.com/pytorch/pytorch/actions/runs/6163427856/job/16727706927))
</p></details>
<details ><summary><b>FLAKY</b> - The following job failed but was likely due to flakiness present on trunk:</summary><p>

* [macos-12-py3-x86-64 / test (default, 1, 4, macos-12)](https://hud.pytorch.org/pr/pytorch/pytorch/107339#16727674082) ([gh](https://github.com/pytorch/pytorch/actions/runs/6163427777/job/16727674082))
</p></details>
<details ><summary><b>UNSTABLE</b> - The following jobs failed but were likely due to flakiness present on trunk and has been marked as unstable:</summary><p>

* [linux-focal-cuda11.8-py3.9-gcc9 / test (multigpu, 1, 1, linux.g5.12xlarge.nvidia.gpu, unstable)](https://hud.pytorch.org/pr/pytorch/pytorch/107339#16727757788) ([gh](https://github.com/pytorch/pytorch/actions/runs/6163427777/job/16727757788))
* [linux-focal-py3.11-clang10 / test (dynamo, 1, 2, linux.2xlarge, unstable)](https://hud.pytorch.org/pr/pytorch/pytorch/107339#16724273623) ([gh](https://github.com/pytorch/pytorch/actions/runs/6162428606/job/16724273623))
* [linux-focal-py3.11-clang10 / test (dynamo, 2, 2, linux.2xlarge, unstable)](https://hud.pytorch.org/pr/pytorch/pytorch/107339#16724273805) ([gh](https://github.com/pytorch/pytorch/actions/runs/6162428606/job/16724273805))
* [linux-focal-py3.8-clang10 / test (dynamo, 1, 2, linux.2xlarge, unstable)](https://hud.pytorch.org/pr/pytorch/pytorch/107339#16724267602) ([gh](https://github.com/pytorch/pytorch/actions/runs/6162428606/job/16724267602))
* [linux-focal-py3.8-clang10 / test (dynamo, 2, 2, linux.2xlarge, unstable)](https://hud.pytorch.org/pr/pytorch/pytorch/107339#16724267789) ([gh](https://github.com/pytorch/pytorch/actions/runs/6162428606/job/16724267789))
</p></details>

This comment was automatically generated by Dr. CI and updates every 15 minutes.

* **Dr.CI #105055**

### :test_tube: See artifacts and rendered test results at [hud.pytorch.org/pr/105055](https://hud.pytorch.org/pr/105055)
* :page_facing_up: Preview [Python docs built from this PR](https://docs-preview.pytorch.org/pytorch/pytorch/105055/index.html)
* :page_facing_up: Preview [C++ docs built from this PR](https://docs-preview.pytorch.org/pytorch/pytorch/105055/cppdocs/index.html)
* :question: Need help or want to give feedback on the CI? Visit the [bot commands wiki](https://github.com/pytorch/pytorch/wiki/Bot-commands) or our [office hours](https://github.com/pytorch/pytorch/wiki/Dev-Infra-Office-Hours)

Note: Links to docs will display an error until the docs builds have been completed.

## :white_check_mark: You can merge normally! (2 Unrelated Failures)
As of commit 1f762dfc92b46323950c3e6e95d99d9687741451 with merge base d0f8ee45bdd3d68895dfecf38b39c363ebf82483 (<sub><sub>image</sub></sub>):
<details ><summary><b>FLAKY</b> - The following job failed but was likely due to flakiness present on trunk:</summary><p>

* [cuda12.1-py3.10-gcc9-sm86-periodic-dynamo-benchmarks / test (aot_eager_torchbench, 1, 1, linux.g5.4xlarge.nvidia.gpu)](https://hud.pytorch.org/pr/pytorch/pytorch/105055#16135412390) ([gh](https://github.com/pytorch/pytorch/actions/runs/5949203435/job/16135412390))
</p></details>
<details ><summary><b>UNSTABLE</b> - The following job failed but was likely due to flakiness present on trunk and has been marked as unstable:</summary><p>

* [linux-focal-rocm5.6-py3.8 / test (default, 1, 3, linux.rocm.gpu, unstable)](https://hud.pytorch.org/pr/pytorch/pytorch/105055#16135257231) ([gh](https://github.com/pytorch/pytorch/actions/runs/5949203290/job/16135257231))
</p></details>

This comment was automatically generated by Dr. CI and updates every 15 minutes.

### CODE REVIEW

File: torchci/lib/jobUtils.ts [hunk 1]
```diff
@@ -144,3 +151,70 @@ export function isSameFailure(
...
+    const matches = trie.search(job.name);
+    if (matches.length <= 1) {
+      // If there is zero or one match, keep the job as it is as this is no retry
+      currentMatch = job;
+    } else {
+      // NB: Default to the latest successful job. This is needed because the event
+      // time from GitHub does not guarantee strict chronological order. A quick
+      // retry event could have a timestamp few seconds earlier than the original
+      // job. We are getting the last successful job here (if any)
+      currentMatch = _.find(matches, (match: T) => isSuccessJob(match));
```
Comment (ZainRizvi) [inline torchci/lib/jobUtils.ts:196]: Why default to the latest successful job instead of simply the latest job, and then seeing if it was successful or not?

This way we'd always show the most recent retry, even in case of failures

File: torchci/lib/jobUtils.ts [hunk 2]
```diff
@@ -144,3 +151,70 @@ export function isSameFailure(
...
+  // The issue here is that the cancelled job name is not the same as the successfully
+  // retried one, for example manywheel-py3_10-cuda11_8-test (cancel) was retried as
+  // manywheel-py3_10-cuda11_8-test / test (success). As their names look different,
+  // HUD and Dr.CI treat them incorrectly as two different jobs and mark the cancelled
+  // one as a failure.
+  //
+  // So the fix here is to check if a cancelled job has been retried successfully and
+  // keep or remove it from the list accordingly.
+  const trie: TrieSearch<T> = new TrieSearch<T>("name", {
+    splitOnRegEx: /\s\/\s/g,
```
Comment (ZainRizvi) [inline torchci/lib/jobUtils.ts:172]: Does this work correctly with sharded jobs? It seems like it might use a successful run of shard 1 to overwrite a failed run on shard 2...

File: torchci/lib/jobUtils.ts [hunk 3]
```diff
@@ -144,3 +151,70 @@ export function isSameFailure(
...
+  //
+  // The issue here is that the cancelled job name is not the same as the successfully
+  // retried one, for example manywheel-py3_10-cuda11_8-test (cancel) was retried as
+  // manywheel-py3_10-cuda11_8-test / test (success). As their names look different,
+  // HUD and Dr.CI treat them incorrectly as two different jobs and mark the cancelled
+  // one as a failure.
+  //
+  // So the fix here is to check if a cancelled job has been retried successfully and
+  // keep or remove it from the list accordingly.
+  const trie: TrieSearch<T> = new TrieSearch<T>("name", {
```
Comment (ZainRizvi) [inline torchci/lib/jobUtils.ts:171]: TIL about trie :D

File: torchci/test/drci.test.ts [hunk 4]
```diff
@@ -29,8 +29,8 @@ export const successfulA = {
 };
 
 const pendingA = {
-  name: "linux-docs / build-docs (cpp)",
-  conclusion: null,
+  name: "linux-docs / build-docs (python)",
```
Comment (ZainRizvi) [inline torchci/test/drci.test.ts:32]: Why was this change needed?

File: torchci/test/jobUtils.test.ts [hunk 5]
```diff
@@ -141,4 +142,133 @@ describe("Test various job utils", () => {
     // Same failure
     expect(isSameFailure(jobA, jobB)).toEqual(true);
   });
+
+  test("test removeCancelledJobAfterRetry", async () => {
```
Comment (ZainRizvi) [inline torchci/test/jobUtils.test.ts:146]: Can you please add a test case for sharded jobs?

**Review by ZainRizvi — CHANGES_REQUESTED** [review]
Looking great for the most part, and really cool to see trie in action.

I'm just concerned about if this might treat sharded jobs as the same job, masking those failures. Can you please confirm that's not the case?

File: torchci/lib/jobUtils.ts (hunk 2 above)
Comment (huydhn) [reply torchci/lib/jobUtils.ts:172]: This should work with shard job because the full name of the job is used.  You could see that the name from Rockset is used as it is without any truncating.  So doing a trie search for `pull / linux-focal-cuda12.1-py3.10-gcc9 / test (default, 1, 5, linux.4xlarge.nvidia.gpu)` in `WORKFLOW NAME / JOB NAME` format will never match with other shards because the shard id `1` is there.

On the other hand, job name like `linux-focal-py3.8-clang10` is actually wrong and only show up when the job is cancelled prematurely.  The correct name should have a suffix like `linux-focal-py3.8-clang10 / build` or `C linux-focal-py3.8-clang10 / test (...)`

In the example, https://hud.pytorch.org/pr/105055 has a bunch of cancel jobs.  If you search for `linux-focal-py3.8-clang10` after the change https://torchci-git-fork-huydhn-remove-wrong-cancel-948947-fbopensource.vercel.app/pr/105055, you can see all the shards are there.

File: torchci/lib/jobUtils.ts (hunk 1 above)
Comment (huydhn) [reply torchci/lib/jobUtils.ts:196]: Well, the issue here is that the latest job according to GitHub timestamp is the one that gets cancelled.  Take jobs from `1f762dfc92b46323950c3e6e95d99d9687741451` as an example, here is what happens for in real time:

1. The `Labeler` job with workflow id `5949201202` at https://github.com/pytorch/pytorch/actions/runs/5949201202 started and was cancelled by the workflow concurrency rule.  Its timestamp was `2023-08-23T08:57:20.619274Z`
2. The successful retried job with id `5949201205` (you can see that the id is bigger) at https://github.com/pytorch/pytorch/actions/runs/5949201205 had the timestamp `2023-08-23T08:57:20.499395Z`

They happened in this order, but you can see that the workflow timestamp of the retry `08:57:20.499395Z` recorded by GitHub was a split second earlier than the cancelled run `08:57:20.619274Z`.

So simply getting the latest job here would wrongly surface the wrong result.  The fix I have here is:

1. Get the timestamp of the latest successful job as the base
2. Check if there are any jobs after that, but at least one minute later.  This is the meaning of `timestamp.isAfter(currentLatestTimestamp, "minute")`
3. Use the conclusion from the latest job found by the second step

File: torchci/test/drci.test.ts (hunk 4 above)
Comment (huydhn) [reply torchci/test/drci.test.ts:32]: Oh, this has the same name as the previous successful mock example `successfulA`.  So if I accidentally include both in my mock `successfulA` and `pendingA`.  Only the pending one is counted.  It's clearer to use another name here I guess.

File: torchci/lib/jobUtils.ts (hunk 2 above)
Comment (huydhn) [reply torchci/lib/jobUtils.ts:172]: I have added a test case to cover this case

**Comment by huydhn** [comment]
> Looking great for the most part, and really cool to see trie in action.
> 
> I'm just concerned about if this might treat sharded jobs as the same job, masking those failures. Can you please confirm that's not the case?

Confirmed that this is working correctly, and I have added a test case to cover this.

**Review by ZainRizvi — APPROVED** [review]

File: torchci/lib/jobUtils.ts (hunk 1 above)
Comment (huydhn) [reply torchci/lib/jobUtils.ts:196]: I don't think the rerun happened after the job succeed. The job was just cancelled and retried so quickly within second, that GitHub timestamps were in the wrong order.  This could happen in a distributed system I think
